File: ruby/tenpin.rb
require "forwardable"

module Bowling
  class Tenpin
    attr_reader :frames

    def initialize
      @frames = Array.new(10) { Frame.new }
    end

    def roll(pinfall)
      frames_handling_roll.each { |f| f.roll(pinfall) }
    end

    def score
      frames.sum(&:score)
    end

    private

    def frames_handling_roll
      Array[frame_handling_normal_roll] + frames_handling_bonus_roll
    end

    def frame_handling_normal_roll
      frames.detect(&:handles_normal_roll?)
    end

    def frames_handling_bonus_roll
      frames.select(&:handles_bonus_roll?)
    end
  end

  class Frame
    extend Forwardable
    def_delegators :@state,
      :score,
      :handles_normal_roll?,
      :handles_bonus_roll?

    attr_reader :rolls
    attr_accessor :state

    def initialize
      @rolls = []
      @state = FrameState.initial_state(self)
    end

    def roll(pinfall)
      rolls << pinfall
      state.transition!
    end

    def strike?
      rolls.first == 10
    end

    def spare?
      !strike? && rolls.take(2).sum == 10
    end

    def open?
      !strike? && !spare?
    end

    def complete?
      ((strike? || spare?) && rolls.length == 3) || (open? && rolls.length == 2)
    end

    def incomplete?
      !complete?
    end
  end

  class FrameState
    def self.initial_state(frame)
      PendingState.new(frame)
    end

    attr_reader :frame

    def initialize(frame)
      @frame = frame
    end

    def transition!
      frame.state = [BonusState, CompleteState, PendingState].detect do |klass|
        klass.state_for?(frame)
      end.new(frame)
    end
  end

  class PendingState < FrameState
    def self.state_for?(frame)
      frame.incomplete?
    end

    def score
      0
    end

    def complete?
      false
    end

    def handles_normal_roll?
      true
    end

    def handles_bonus_roll?
      false
    end
  end

  class BonusState < FrameState
    def self.state_for?(frame)
      (frame.strike? || frame.spare?) && frame.incomplete?
    end

    def score
      0
    end

    def handles_normal_roll?
      false
    end

    def handles_bonus_roll?
      true
    end
  end

  class CompleteState < FrameState
    def self.state_for?(frame)
      frame.complete?
    end

    def score
      frame.rolls.sum
    end

    def complete?
      true
    end

    def handles_normal_roll?
      false
    end

    def handles_bonus_roll?
      false
    end
  end
end

require "minitest/autorun"

class TestTenpin < Minitest::Test
  def setup
    @game = Bowling::Tenpin.new
  end

  def test_initial_score
    assert_equal @game.score, 0
  end

  def test_complete_open_frame
    2.times { @game.roll(1) }
    assert_equal 2, @game.score
  end

  def test_multiple_complete_open_frames
    10.times { @game.roll(1) }
    assert_equal 10, @game.score
  end

  def test_incomplete_open_frame
    @game.roll(1)
    assert_equal 0, @game.score
  end

  def test_multiple_open_frames_with_last_incomplete
    5.times { @game.roll(1) }
    assert_equal 4, @game.score
  end

  def test_all_gutter_balls
    20.times { @game.roll(0) }
    assert_equal 0, @game.score
  end

  def test_complete_strike_followed_by_complete_open_frame
    [10,1,1].each { |pinfall| @game.roll(pinfall) }
    assert_equal 14, @game.score
  end
end

class TestFrame < Minitest::Test
  def setup
    @frame = Bowling::Frame.new
  end

  def test_incomplete_open_frame
    @frame.roll(1)
    assert_equal 0, @frame.score
  end

  def test_complete_open_frame
    2.times { @frame.roll(1) }
    assert_equal 2, @frame.score
  end

  def test_strike
    @frame.roll(10)
    assert @frame.strike?
  end

  def test_spare
    2.times { @frame.roll(5) }
    assert @frame.spare?
  end

  def test_open
    2.times { @frame.roll(1) }
    assert @frame.open?
  end

  def test_complete_strike
    [10,1,1].each { |pinfall| @frame.roll(pinfall) }
    assert @frame.complete?
  end

  def test_complete_spare
    [5,5,1].each { |pinfall| @frame.roll(pinfall) }
    assert @frame.complete?
  end

  def test_complete_open
    2.times { @frame.roll(1) }
    assert @frame.complete?
  end

  def test_incomplete_strike
    [10,1].each { |pinfall| @frame.roll(pinfall) }
    assert @frame.incomplete?
  end

  def test_incomplete_spare
    [8,2].each { |pinfall| @frame.roll(pinfall) }
    assert @frame.incomplete?
 end

  def test_incomplete_open
    @frame.roll(9)
    assert @frame.incomplete?
  end
end

File: elixir/tenpin.exs
defmodule Bowling.Tenpin do
  # strike
  def score([10, bonus1, bonus2 | tail]) do
    10 + bonus1 + bonus2 + score([bonus1, bonus2 | tail])
  end

  # spare
  def score([spare1, spare2, bonus | tail]) when spare1 + spare2 == 10 do
    spare1 + spare2 + bonus + score([bonus | tail])
  end

  # open frame
  def score([roll1, roll2 | tail]) when roll1 + roll2 < 10 do
    roll1 + roll2 + score(tail)
  end

  # fallthrough
  def score([_ | _]), do: 0 # incomplete/unscoreable frame
  def score([]), do: 0 # no rolls left
end

ExUnit.start()

defmodule TenpinTest do
  use ExUnit.Case
  alias Bowling.Tenpin, as: Tenpin

  test "initial score" do
    assert Tenpin.score([]) == 0
  end

  test "complete open frame" do
    assert Tenpin.score([1,1]) == 2
  end

  test "multiple complete open frames" do
    assert Tenpin.score([1,1,1,1,1,1,1,1,1,1]) == 10
  end

  test "incomplete open frame" do
    assert Tenpin.score([1]) == 0
  end

  test "multiple open frames with last incomplete" do
    assert Tenpin.score([1,1,1,1,1]) == 4
  end

  test "all gutter balls" do
    rolls = [0,0,0,0,0,0,0,0,0,0]
    assert Tenpin.score(rolls) == 0
  end

  test "complete strike followed by complete open frame" do
    assert Tenpin.score([10,1,1]) == 14
  end

  test "complete strike followed by incomplete spare" do
    assert Tenpin.score([10,5,5]) == 20
  end

  test "complete strike followed by complete spare" do
    assert Tenpin.score([10,5,5,1]) == 31
  end

  test "incomplete strike, no bonus rolls" do
    assert Tenpin.score([10]) == 0
  end

  test "incomplete strike, one bonus roll" do
    assert Tenpin.score([10,1]) == 0
  end

  test "perfect game" do
    rolls = [10,10,10,10,10,10,10,10,10,10,10,10]
    assert Tenpin.score(rolls) == 300
  end

  test "complete spare" do
    assert Tenpin.score([3,7,4]) == 14
  end

  test "complete spare followed by complete open frame" do
    assert Tenpin.score([3,7,4,1]) == 19
  end

  test "complete spare followed by strike with no bonus rolls" do
    assert Tenpin.score([3,7,10]) == 20
  end

  test "complete spare followed by strike with one bonus roll" do
    assert Tenpin.score([3,7,10,1]) == 20
  end

  test "all fives" do
    rolls = [5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]
    assert Tenpin.score(rolls) == 150
  end
end
